### PR TITLE
Adds Dragon Darts effect (CFRU port)

### LINKED ISSUE #1058: MOVE_DRAGON_DARTS Effect

### PR DESCRIPTION
**Pull request by AlexOn1ine**
## Description
Dragon Darts port from the CFRU. 

Fixed all issues from #2842, addressed mgriffins review comments and created `ABILITYEFFECT_WOULD_BLOCK` and `ABILITYEFFECT_WOULD_ABSORB`as recommended. 

Also left two commented out tests because they are illegal. I had them for testing purposes. I can remove them after the review. 

## Issue(s) that this PR fixes
Closes  #1058
Supersedes #2842

## **People who collaborated with me in this PR**
@Skeli789 for the original code 
@BuffelSaft for the port from the CFRU

### CODE REVIEW

**Comment by BuffelSaft** [comment]
Something worth testing (but which shouldn't block the PR because it's a silly edge case):
What happens when Dragon Darts targets an ally, but the ally uses Ally Switch first? In Sword this causes the darts to target and hit the user. Would be a good test of how robust the smart targeting is.

**Comment by AlexOn1ine** [comment]
> Something worth testing (but which shouldn't block the PR because it's a silly edge case): What happens when Dragon Darts targets an ally, but the ally uses Ally Switch first? In Sword this causes the darts to target and hit the user. Would be a good test of how robust the smart targeting is.

I'm not sure what is supposed to happen but here is a recording:

https://github.com/rh-hideout/pokeemerald-expansion/assets/93446519/ec07b6ef-51d7-4ae2-8705-473de015eaf4

**Comment by AlexOn1ine** [comment]
> > I'm not sure what is supposed to happen (I guess it is according to Sword and Shield) but here is a recording:
> > ally_switch.mp4
> 
> Just tested in Sword with an Ally Switch Togekiss. In that case it makes both darts hit Dragapult so it KOs itself (I guess because Togekiss is immune?), so that looks good!

yep, this is what happens

https://github.com/rh-hideout/pokeemerald-expansion/assets/93446519/35c344a7-aeec-4eac-8d9b-7edcce81e2ae

File: test/battle/move_effect/dragon_darts.c [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+    } SCENE {
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_DRAGON_DARTS, playerLeft);
+        HP_BAR(opponentRight);
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_DRAGON_DARTS, playerLeft);
+        HP_BAR(opponentRight);
+        MESSAGE("Hit 2 time(s)!");
+    }
+}
+
+// DOUBLE_BATTLE_TEST("Fire Dragon Darts")
```
Comment (Bassoonian) [inline test/battle/move_effect/dragon_darts.c:191]: Is there a reason these tests are commented out?

File: src/battle_util.c [hunk 2]
```diff
@@ -11001,13 +11026,13 @@ void TryRestoreHeldItems(void)
...
+            if (ItemId_GetPocket(lostItem) == POCKET_BERRIES && GetMonData(&gPlayerParty[i], MON_DATA_HELD_ITEM) != lostItem)
                 lostItem = ITEM_NONE;
 
             // Check if the lost item should be restored
-            if ((lostItem != ITEM_NONE || returnNPCItems) && ItemId_GetPocket(lostItem) != POCKET_BERRIES) 
+            if ((lostItem != ITEM_NONE || returnNPCItems) && ItemId_GetPocket(lostItem) != POCKET_BERRIES)
                 SetMonData(&gPlayerParty[i], MON_DATA_HELD_ITEM, &lostItem);
-            
+
         }
```
Comment (Bassoonian) [inline src/battle_util.c:11036]:
```suggestion
        }
```

File: src/battle_util.c [hunk 3]
```diff
@@ -5197,6 +5199,29 @@ u32 AbilityBattleEffects(u32 caseID, u32 battler, u32 ability, u32 special, u32
...
+                    effect = 3;
+                }
+                else
+                {
+                    gBattleCommunication[MULTISTRING_CHOOSER] = B_MSG_FLASH_FIRE_NO_BOOST;
+                    if (gProtectStructs[gBattlerAttacker].notFirstStrike)
+                        gBattlescriptCurrInstr = BattleScript_FlashFireBoost;
+                    else
+                        gBattlescriptCurrInstr = BattleScript_FlashFireBoost_PPLoss;
+
```
Comment (Bassoonian) [inline src/battle_util.c:5222]:
```suggestion
                        gBattlescriptCurrInstr = BattleScript_FlashFireBoost_PPLoss;
```

File: test/battle/move_effect/dragon_darts.c (hunk 1 above)
Comment (AlexOn1ine) [reply test/battle/move_effect/dragon_darts.c:191]: There for testing purposes. I remove them after the review.

File: src/battle_util.c (hunk 3 above)
Comment (AlexOn1ine) [reply src/battle_util.c:5222]: what's the change?

Comment (Bassoonian) [reply src/battle_util.c:5222]: Oh, thanks GitHub. It's the redundant newline after

**Comment by AlexOn1ine** [comment]
Reviews addressed. Tests removed.

**Review by Bassoonian — APPROVED** [review]
